Generic generate methods

bmen25124 opened the pull request:
Adding preset support to payload generation methods. So extensions could pick a chat/text preset and generate payload according to that.

## Checklist:

- [x] I have read the [Contributing guidelines](https://github.com/SillyTavern/SillyTavern/blob/release/CONTRIBUTING.md).

bmen25124 committed:
sendOpenAIRequest/getTextGenGenerationData methods are improved, now it can use custom API, instead of active ones

bmen25124 commented on public/scripts/openai.js:
```diff
@@ -224,6 +224,87 @@ const sensitiveFields = [
+    reasoning_effort: ['#openai_reasoning_effort', 'reasoning_effort', false],
+    seed: ['#seed_openai', 'seed', false],
+    n: ['#n_openai', 'n', false],
+    jailbreak_system: ['#jailbreak_system', 'jailbreak_system', false],
+    bypass_status_check: ['#openai_bypass_status_check', 'bypass_status_check', false],
```
I added these 2 lines. Why did they not exist before?

Cohee1207 replied:
> jailbreak_system

Dead field, doesn't exist anymore

> bypass_status_check

This one was missed.

bmen25124 committed:
Added missing model param

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -1835,16 +1921,43 @@ async function sendAltScaleRequest(messages, logit_bias, signal, type) {
+    // Merge with oai_settings
+    const resultKeys = new Set(Object.keys(result));
+    for (const [key, value] of Object.entries(oai_settings)) {
+        if (!resultKeys.has(key)) {
```
```suggestion
        if (!Object.hasOwn(result, key)) {
```

bmen25124 replied:
Fixed. By the way, I'm not sure about merging strategy. I mean, we had to merge with `oai_settings` but what if I want to merge fields that are not in `settingsToUpdate`? In the future, we could change this method. Maybe add with `strategy` parameter.

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -1857,39 +1970,42 @@ async function sendOpenAIRequest(type, messages, signal) {
 
     messages = messages.filter(msg => msg && typeof msg === 'object');
 
+    const active_oai_settings = structuredClone(preset ? presetToSettings(preset) : oai_settings);
```
Just `settings` will be nicer

bmen25124 replied:
Fixed

Cohee1207 commented on public/scripts/textgen-settings.js:
```diff
@@ -402,10 +402,12 @@ function getTokenizerForTokenIds() {
  */
-function getCustomTokenBans() {
-    if (!settings.send_banned_tokens || (!settings.banned_tokens && !settings.global_banned_tokens && !textgenerationwebui_banned_in_macros.length)) {
+function getCustomTokenBans(bannedTokens = null) {
```
why not pass a settings object? There's also `global_banned_tokens`

bmen25124 replied:
I could, but we are only using `settings.banned_tokens ` because `global_banned_tokens ` is not a preset field. If you want it, I can do it, there is no problem for me.

Cohee1207 replied:
I think passing a full object is better since you could easily extend this in the future. And use object destructuring in function definition to basically have a local variable of `banned_tokens`.

`function getCustomTokenBans({ banned_tokens })`

Cohee1207 replied:
Also it generally shouldn't be a concern of this code to distinguish between preset and non-preset settings as it accepts settings object, not a preset object. It's just so happens that for TC the names match.

bmen25124 replied:
Fixed. 

We can do the same thing for others in the future. Like `getTextGenServer`, `getTextGenModel`

bmen25124 committed:
Removed unnecessary variable

bmen25124 committed:
active_oai_settings -> settings

bmen25124 committed:
Merge branch 'staging' into generic_generate_methods

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -1835,16 +1921,42 @@ async function sendAltScaleRequest(messages, logit_bias, signal, type) {
     return data.output;
 }
 
+export function presetToSettings(preset) {
```
You have two functions named `presetToSettings` it would be fine, but both are exported. Please rename them to be different names `textCompletionPresetToSettings` and `chatCompletionPresetToSettings` are my best contenders

bmen25124 replied:
Fixed

Cohee1207 commented on public/scripts/textgen-settings.js:
```diff
@@ -1263,168 +1278,206 @@ export function replaceMacrosInList(str) {
+ * @returns
+ */
+export async function getTextGenGenerationData(finalPrompt, maxTokens, isImpersonate, isContinue, cfgValues, type, { preset, model, textgen_type, api_url } = {}, emitEvent = true) {
+    const active_settings = structuredClone(preset ? presetToSettings(preset) : settings);
```
It would be cooler to leave this local variable name as just `settings` as in openai.js, but it then collides with a module-scoped variable of the same name. So my proposal is to renamed the module-scoped `settings` to be the same as its exported name of `textgenerationwebui_settings` and leave it as `settings` here. It would be more consistent with CC.

bmen25124 replied:
Good idea, fixed.

Cohee1207 replied:
Thank you my good man

bmen25124 committed:
settings -> textgenerationwebui_settings

bmen25124 committed:
Better presetToSettings names, simpler settings name in getTextGenGenerationData,

bmen25124 committed:
Removed unused jailbreak_system

Cohee1207 reviewed (changes requested):
These functions which are called from sendOpenAIRequest are still using the global oai_settings object:
* `calculateLogitBias`
* `parseChatCompletionLogprobs`
* `sendAltScaleRequest`
* `isToolCallingSupported`
* `getStreamingReply`

Same story in getTextGenGenerationData but simpler:
* `calculateLogitBias`
* `getLogprobsNumber` (not called with type argument)

There's probably more. These are just those that I found from a quick glance.
I'm afraid we have to rethink the chosen approach. Too many side effects and possible "could go horribly wrong" places.

bmen25124 committed:
Reverted most core changes, new custom-request.js file

bmen25124 commented:
@Cohee1207 I reverted the most changes and added new classes/methods `custom-request.js` file. What do you think so far?

* Names can be changed
* New file looks like an extension code because it contains very little from the core. So why not I'm creating a base extension? Because:
    - `requires` not working for extension.
    - For general availability, it is better if it were in the core. If we are going to keep simple, I don't think there is an important tech debt or maintenance for this code.
* If we're going to merge this, we had to squash. Because commits are already messy.

Cohee1207 commented on public/scripts/custom-request.js:
```diff
@@ -0,0 +1,159 @@
+     * @param {string} type
+     * @returns {string} API URL
+     */
+    static getApiUrl(type) {
```
This function either has to be reused with textgen-settings (used there) or imported. This one is perfectly fine to be reused as it's a plain switch, but I'm worried with updating it in two places when adding new sources.

bmen25124 committed:
Forced stream to false, removed duplicate method, exported settingsToUpdate

Cohee1207 committed:
Rewrite typedefs to define props one by one

bmen25124 committed:
Added extractData param for simplicity

bmen25124 committed:
Merge branch 'generic_generate_methods' of github.com:bmen25124/SillyTavern into generic_generate_methods

bmen25124 committed:
Fixed typehints

bmen25124 committed:
Fixed typehints (again)
